Implement urgency base functionality

RyanDwyer opened the pull request:
Introduces a command to manually set urgency, as well as rendering of urgent views, sending the IPC event, removing urgency one second after focus, and matching urgent views via criteria.

To test, run something like `swaymsg [class="Firefox"] urgent enable`.

Try it with Firefox on a different workspace too, and notice the swaybar indicator working.

Note that urgency is not yet implemented when xwayland views request it, nor when a view requests focus (depending on the value of `focus_on_window_activation`). I feel like that's out of scope for this PR.

Closes #2258.

ianyfan commented on sway/tree/view.c:
```diff
@@ -1047,3 +1053,27 @@ bool view_is_visible(struct sway_view *view) {
+	container_damage_whole(view->swayc);
+
+	struct sway_container *ws = container_parent(view->swayc, C_WORKSPACE);
+	ipc_event_workspace(ws, NULL, "urgent");
```
The arguments should be `NULL, ws, "urgent"` since the prototype has the old workspace first, then the current ("new") workspace. Also, do you send an event for the window too?

SirCmpwn commented:
Overall this LGTM

RyanDwyer commented:
I just realised that implementing the xwayland urgency hint was really easy, so I went and did that.

I also had a look at `focus_on_window_activation` but I think some changes would need to be made in wlroots (eg. a `request_activate` event).

RedSoxFan commented:
1. Splitting is working again after your last commit (woot)
2. I can't seem to get `urgent enable` to work for any floating view (it works fine for tiling)
3. If you close a window without focusing it while it's urgency flag is set, the workspace retains the urgent flag.
4. Should `urgent deny` be extended to prevent `urgent enable` or even something like `urgent block` added as a stricter option?
5. Should the urgency flag be propagated upwards for nested containers? I'd have to check i3's behavior on this

RyanDwyer commented:
> 2. I can't seem to get urgent enable to work for any floating view (it works fine for tiling)

Fixed. This was a problem affecting all criteria. `container_for_each_descendant_dfs` did not descend into floating views. It does now.

> 3. If you close a window without focusing it while it's urgency flag is set, the workspace retains the urgent flag.

Fixed. The same issue occurred when moving urgent views to new workspaces. When reviewing this PR, start by looking at `workspace_detect_urgent`. The rest is mostly calling that function in various places.

> 4. Should urgent deny be extended to prevent urgent enable or even something like urgent block added as a stricter option?

I don't think so. The `deny` setting is there to prevent a noisy client from claiming unwanted urgency. `swaymsg urgent enable` is only run by the user or via a script of their's and should not be affected by the `deny` setting. 

> 5. Should the urgency flag be propagated upwards for nested containers? I'd have to check i3's behavior on this

I can't check this right now, so I'll either wait for your response or I'll check it later.

RedSoxFan commented:
> > 5. Should the urgency flag be propagated upwards for nested containers? I'd have to check i3's behavior on this
>
> I can't check this right now, so I'll either wait for your response or I'll check it later.

Yes, i3 does propagate the urgent flag up the tree.

--

Aside from the urgency propagation, this LGTM.

ascent12 commented on sway/criteria.c:
```diff
@@ -46,6 +46,31 @@ static int regex_cmp(const char *item, const pcre *regex) {
 }
 
+static int cmp_urgent(const void *_a, const void *_b) {
+	struct sway_view *a = *(void **)_a;
```
Minor point, but don't cast away the const here.

RyanDwyer replied:
I copied this from `container_sort_workspaces`. When I tried it without the cast something weird was happening and it wasn't comparing them properly.

RedSoxFan commented:
Just needs the urgent border colors set when there is an urgent descendant and this should be good

SirCmpwn commented:
There's an off-by-one error somewhere

![](https://sr.ht/kuhU.png)

SirCmpwn commented:
Actually I think it's because the urgent colors aren't correctly pulled out of the bar config in `ipc.c`

RyanDwyer commented:
Fixed. Swaybar was ignoring what IPC gave it and was using default colours. This would only be noticeable if you had an `urgent_workspace_border` colour that differed from the default.

SirCmpwn commented:
Sweet.
